wnfs/src/private/directory.rs

```diff
@@ -287,23 +291,64 @@ impl PrivateDirectory {
         }
     }
 
+    /// This should be called to prepare a node for modifications,
+    /// if it's meant to be a successor revision of the current revision.
+    ///
+    /// It will store the current revision in the given `BlockStore` to
+    /// retrieve its CID and put that into the `previous` links,
+    /// as well as advancing the ratchet and resetting the `persisted_as` pointer.
+    pub(crate) async fn prepare_next_revision<B: BlockStore>(
+        self: Rc<Self>,
+        store: &mut B,
+        rng: &mut impl RngCore,
+    ) -> Result<Self> {
+        let cid = self.store(store, rng).await?;
+
+        let mut cloned = Rc::try_unwrap(self).unwrap_or_else(|rc| (*rc).clone());
+        cloned.persisted_as = OnceCell::new(); // Also done in `.clone()`, but need this to work in case try_unwrap optimizes.
+        let key = cloned.header.get_private_ref().revision_key.0;
+        let previous = Encrypted::from_value(BTreeSet::from([cid]), &key, rng)?;
+
+        cloned.previous = Some(previous);
+        cloned.header.advance_ratchet();
+
+        Ok(cloned)
+    }
+
+    /// This prepares this file for key rotation, usually for moving or
+    /// copying the file to some other place.
+    ///
+    /// Will reset the ratchet, so a different key is necessary for read access,
+    /// will reset the inumber to reset write access,
+    /// will update the bare namefilter to match the new parent's namefilter,
+    /// so it inherits the write access rules from the new parent and
+    /// resets the `persisted_as` pointer.
+    pub(crate) fn prepare_key_rotation(
```

Comment:
How do we handle deterministic keys here? Is there a need to expose an alternative `attach` API for allowing deterministic key rotation?

Reply:
That's a good question.

Generally, there's not *really* such a thing as "deterministic key rotation". When you rotate keys, you want to make sure that anyone who could read *previous* versions, won't be able to read *future* versions.
If you make that process deterministic however, then anyone who knows the past will be able to just apply the deterministic operation themselves to be able to read the present/future.

That said, one could argue our `Ratchet` is sort-of "deterministic key rotation": We constantly 'rotate' the key for every revision.

Anyway.

I think once we have shared private files, we should talk to Functionland again. Perhaps it's better to not generate `PrivateDirectory` and `PrivateRef` deterministically, but instead generate a single exchange public key deterministically, and 'share' read access of the public directory with that exchange key.